### DISCUSSION TITLE
kubectl port-forward bypass sidecar interception

### DISCUSSION THREAD

**Discussion by gabriel9895** [istio/istio#48994] [Q&A] [answered]
When testing with istio and bookinfo applications, I found out that after```kubectl port-forward svc/productpage -n default 9080```, the traffic from host machine via kubectl port-forward bypass istio iptables interception and didn't go through envoyproxy. So I want to ask the reason behind it.

The facts we found:
First [have strict MTLS over the mesh](https://istio.io/latest/docs/tasks/security/authentication/mtls-migration/#lock-down-mutual-tls-for-the-entire-mesh), then 1. we made HTTP request from sleep pod without sidecar to bookinfo app in mesh, the traffic failed due to strict MTLS as expected. 2. We ``kubectl port-forward svc/productpage -n default 9080``` and curl localhost:9080/productpage from host where bookinfo is deployed on. But we see traffic is succeed.
Second, we verified that indeed traffic didn't even reach envoyproxy via bpftrace scripts [here](https://blog.mygraphql.com/zh/posts/low-tec/trace/trace-istio/trace-istio-part3/#3-%E6%9F%A5%E7%9C%8B%E8%BE%93%E5%87%BA%E8%B7%9F%E8%B8%AA%E7%BB%93%E6%9E%9C), basically the script tracks the TCP connection and listener filter stack creation, so as long as TCP connection established with 15006 and ListenerFilterChain starts to be created, bpftrace will capture the syscall. But in the port-forward case, nothing is printed, meaning envoyproxy is not involved. Further I verified with tcpdump inside productpage pod, there is only traffic among 127.0.0.1 to itself.(I attached dump via portforward and sleep pod(10.244.0.168) productpage(10.244.0.174))  [pcap.zip](https://github.com/istio/istio/files/14053949/pcap.zip)

It confuses us:
what happens to traffic from kubectl portforward? Do they skip overall the iptables interception or they exits early(e.g. PREROUTING -> FORWARD, not returned INPUT)? 
From my understanding on [istio iptables](https://github.com/istio/istio/wiki/Understanding-IPTables-snapshot#use-pid-to-get-iptables), all TCP traffic (portforward under SDPY included?) will be hackjacked from PREROUTING -> ISTIO_INBOUND -> as we are port 9080 ISTIO_IN_REDIRECT -> port 15006 -> envoyproxy virtualinbound handler.
```
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A ISTIO_INBOUND -p tcp -m tcp --dport 22 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15090 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15020 -j RETURN
-A ISTIO_INBOUND -p tcp -j ISTIO_IN_REDIRECT
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
```
Also as @rootsongjc explained in https://tetrate.io/blog/traffic-types-and-iptables-rules-in-istio-sidecar-explained/ I thought port-forward belongs to Type 1: Remote Pod -> Local Pod, but fact seems not.
In this case, I don't see how could localhost:9080 from port-forward be an exception to skip this flow chain? OR port-forward mechanism doesn't involve envoyproxy at all and directly contacts productpage container, how?

Therefore we would like to ask for support and explanation from istio experts! Thank you very much!

**Comment by howardjohn** [comment]
https://stuffphilwrites.com/wp-content/uploads/2014/09/FW-IDS-iptables-Flowchart-v2019-04-30-1.png - see "localhost source" under incoming packet. nat PREROUTING is never hit

**Comment by rootsongjc** [comment]
The client access flows like this: Client -> localhost -> Kubernetes API Server -> Server Pod, so for the Server Pod to see requests from localhost, they don't go through the sidecar.

It also reminded me that requests inside the pod will bypass Envoy too. The communication only happens in the Linux user namespace. So the diagram of Type4 in this [blog](https://tetrate.io/blog/traffic-types-and-iptables-rules-in-istio-sidecar-explained/) is a bit wrong. I will fix that.

**Comment by gabriel9895** [comment]
Thank you both very much for the explanation. Yes, I understand that for the pod container to see **internal** requests from localhost w.r.t the pod network namespace, the NAT table in the prerouting chain will not be touched as @howardjohn  stated.
But for the port forward from localhost w.r.t the host machine, traffic flows from the locahost to the API server to the app service, then the service directs the traffic to the app pod. Then in the view of the app pod, isn't the request with **svc IP:9080**? Then in the view of the pod network namespace iptables, why the request is still considered as "localhost source", as localhost is in the context of my host machine?

To compare, I did the experiment using ```kubectl port-forward --address=100.79.51.198 svc/productpage 9080``` or even ```kubectl port-forward --address=0.0.0.0 svc/productpage 9080```, where ```100.79.51.198``` is my host eth0 IP addr. But when requesting locally from host ```curl 100.79.51.198:9080/productpage```, the traffic still managed to bypass the envoyproxy, which I don't understand the reason behind. How do the iptables in the pod network namespace interpret the "localhost"?

**Reply by howardjohn** [reply, marked as answer]
One note, _service_ is not involved in port forward. When you port forward to svc/productpage, it just picks a random pod in productpage svc.

When you portforward, the request is not coming from these external addresses. Its more align with doing something like `kubectl exec POD -- curl ...`. The request comes from inside the pod network namespace. Its not just a raw request to the pod IP from external

**Reply by gabriel9895** [reply]
Thanks a lot @howardjohn  for the explanation! It's now clear why the port forward requests is not captured by the envoyproxy :)
